lib/rtti.h

```diff
@@ -0,0 +1,181 @@
+/*
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+#ifndef LIB_RTTI_H_
+#define LIB_RTTI_H_
+
+#include <algorithm>
+#include <array>
+#include <cstdint>
+#include <type_traits>
+
+namespace RTTI {
+
+using TypeId = std::uint64_t;
+static constexpr TypeId InvalidTypeId = UINT64_C(0);
+
+namespace detail {
+// Apparently string_view does not optimize properly in constexpr context
+// for gcc < 13
+struct TypeNameHolder {
+    template <size_t N>
+    constexpr TypeNameHolder(const char (&str)[N]) : str(&str[0]), length(N - 1) {}  // NOLINT
+
+    const char *str;
+    size_t length;
+};
+
+template <typename T>
+constexpr TypeNameHolder TypeNameProxy() {
+#ifdef __clang__
+    return __PRETTY_FUNCTION__;
+#elif defined(__GNUC__)
+    return __PRETTY_FUNCTION__;
+#else
+#error "Unsupported compiler"
+#endif
+}
+
+static constexpr uint64_t FNV1a(const char *str, size_t n,
+                                uint64_t hash = UINT64_C(0xcbf29ce484222325)) {
+    return n == 0 ? hash : FNV1a(str + 1, n - 1, UINT64_C(0x100000001b3) * (hash ^ str[0]));
+}
+
+static constexpr uint64_t FNV1a(TypeNameHolder str) { return FNV1a(str.str, str.length); }
```

Comment:
Nitpick: it would be cleaner to write this as a `for` loop and not have the `const char *` version at all.

Reply:
I see. I did not know the `constexpr` implementation according to the C++17 rules is still so buggy.

Reply:
Yeah. `constexpr`'s are quite hard from the compiler perspective. Also, they put some additional pressure on standard library and they allowed by a standard to fallback to runtime implementation. So yes, we need to ensure that the generated code is what we'd expect. C+20 will simplify lots of things here.